lib/commons/dom/get-text-element-stack.js

```diff
@@ -38,14 +38,29 @@ function getTextElementStack(node) {
       range.selectNodeContents(elm);
       const rects = range.getClientRects();
 
-      // if the elements bounding rect is smaller than the text rects
-      // (for example, it's using text truncation or overflow) we
-      // need to get the rect of the element rather than look at the
-      // text node rects as they will return the full width of the
-      // text node (which can go off the screen)
+      // if any text rect is larger than the bounds of the parent,
+      // or goes outside of the bounds of the parent, we need to use
+      // the parent rect so we stay within the bounds of the element.
+      //
+      // since we use the midpoint of the element when determining
+      // the rect stack we will also use the midpoint of the text rect
+      // to determine out of bounds.
+      //
       // @see https://github.com/dequelabs/axe-core/issues/2178
       // @see https://github.com/dequelabs/axe-core/issues/2483
-      if (Array.from(rects).some(rect => rect.width > nodeRect.width)) {
+      // @see https://github.com/dequelabs/axe-core/issues/2681
+      const outsideRectBounds = Array.from(rects).some(rect => {
+        const horizontalMidpoint = rect.left + rect.width / 2;
+        const verticalMidpoint = rect.top + rect.height / 2;
+
+        return (
+          horizontalMidpoint < nodeRect.left ||
+          horizontalMidpoint > nodeRect.left + nodeRect.width ||
+          verticalMidpoint < nodeRect.top ||
+          verticalMidpoint > nodeRect.top + nodeRect.height
```

Comment:
I think this could be:

```suggestion
          horizontalMidpoint < nodeRect.left ||
          horizontalMidpoint > nodeRect.right ||
          verticalMidpoint < nodeRect.top ||
          verticalMidpoint > nodeRect.bottom
```